openshift/catalog-sync/pkg/extract/extract.go

```diff
@@ -0,0 +1,184 @@
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("oci-%s-", name))
+	if err != nil {
+		return nil, fmt.Errorf("create temp dir: %w", err)
+	}
+
+	cleanup := func() { os.RemoveAll(tmpDir) }
+
+	if err := pullImageToLayout(ctx, imageRef, tmpDir); err != nil {
+		cleanup()
+		return nil, fmt.Errorf("pull image: %w", err)
+	}
+
+	if err := extractFilesystem(ctx, tmpDir); err != nil {
+		cleanup()
+		return nil, fmt.Errorf("extract filesystem: %w", err)
+	}
+
+	store, err := oci.New(tmpDir)
+	if err != nil {
+		cleanup()
+		return nil, fmt.Errorf("open OCI store: %w", err)
+	}
+
+	return &Result{
+		Store:   store,
+		TmpDir:  tmpDir,
+		Cleanup: cleanup,
+	}, nil
+}
+
+// pullImageToLayout pulls an image and writes it to OCI layout at the specified directory
+func pullImageToLayout(ctx context.Context, imageRef, tmpDir string) error {
+	srcRef, err := docker.ParseReference("//" + imageRef)
+	if err != nil {
+		return fmt.Errorf("parse image ref: %w", err)
+	}
+
+	destRef, err := layout.ParseReference(fmt.Sprintf("%s:%s", tmpDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout ref: %w", err)
+	}
+
+	policy := &signature.Policy{
+		Default: []signature.PolicyRequirement{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	}
+	policyCtx, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+	defer policyCtx.Destroy()
+
+	// Required to allow run the tests on Mac
+	sysCtx := &types.SystemContext{
+		OSChoice:           "linux",
+		ArchitectureChoice: "amd64",
+	}
+
+	if _, err := imgcopy.Image(ctx, policyCtx, destRef, srcRef, &imgcopy.Options{
+		SourceCtx:      sysCtx,
+		DestinationCtx: sysCtx,
+	}); err != nil {
+		return fmt.Errorf("copy image to layout: %w", err)
+	}
+
+	return nil
+}
+
+// extractFilesystem extracts the filesystem layers from the OCI image into <tmpDir>/fs
+func extractFilesystem(ctx context.Context, ociDir string) error {
+	ref, err := layout.ParseReference(fmt.Sprintf("%s:%s", ociDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout: %w", err)
+	}
+	src, err := ref.NewImageSource(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("open image source: %w", err)
+	}
+	defer src.Close()
+
+	manifestBytes, _, err := src.GetManifest(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("get manifest: %w", err)
+	}
+
+	mf, err := manifest.FromBlob(manifestBytes, manifest.GuessMIMEType(manifestBytes))
+	if err != nil {
+		return fmt.Errorf("parse manifest: %w", err)
+	}
+
+	fsDir := filepath.Join(ociDir, "fs")
+	if err := os.MkdirAll(fsDir, 0755); err != nil {
+		return fmt.Errorf("make fs dir: %w", err)
+	}
+
+	for _, layer := range mf.LayerInfos() {
+		rc, _, err := src.GetBlob(ctx, layer.BlobInfo, nil)
+		if err != nil {
+			return fmt.Errorf("get layer: %w", err)
+		}
+		if err := untarLayer(rc, fsDir); err != nil {
+			rc.Close()
+			return fmt.Errorf("extract layer: %w", err)
+		}
```

Comment:
We should use an existing libraries that are designed specifically for extracting image layers here.
- For automatic decompression: https://github.com/containers/image/blob/e2534e0da6eced91be073195189f2394ab72923a/pkg/compression/compression.go#L160
- For applying layers correctly: perhaps https://github.com/containers/storage/blob/29f052f570bb3c2839a55448faeb792dd03ac90d/pkg/archive/archive.go#L1046. Another option is https://github.com/containerd/containerd/blob/1e8955d8b7c5ee9ec6c2cdb1e6eedff0c7088f9d/pkg/archive/tar.go#L141

Reply:
I will give a look on that. 👍

Reply:
Changed here to simplify